### PR TITLE
IntervalShardingAlgorithm formats for LocalDateTime and Date

### LINKED ISSUE #16738: `IntervalShardingAlgorithm` uses `toString()` method on `LocalDateTime` resulting in `StringIndexOutOfBoundsException`
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?

- 5.1.0

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

- ShardingSphere-JDBC

### Expected behavior

- The following unit tests work fine.
```java
import com.google.common.collect.Range;
import org.apache.shardingsphere.infra.datanode.DataNodeInfo;
import org.apache.shardingsphere.sharding.api.sharding.standard.PreciseShardingValue;
import org.apache.shardingsphere.sharding.api.sharding.standard.RangeShardingValue;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.LongStream;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertThat;

@Test
public void test1() {
    LocalDate startTime = LocalDate.of(2022, 2, 16);
    LocalDate endTime = LocalDate.of(2022, 4, 11);
    IntervalShardingAlgorithm intervalShardingAlgorithm = new IntervalShardingAlgorithm();
    intervalShardingAlgorithm.getProps().setProperty("datetime-pattern", "yyyy-MM-dd HH:mm:ss.SSS");
    intervalShardingAlgorithm.getProps().setProperty("datetime-lower", "2022-02-16 00:00:00.000");
    intervalShardingAlgorithm.getProps().setProperty("datetime-upper", "2022-04-11 23:59:59.999");
    intervalShardingAlgorithm.getProps().setProperty("sharding-suffix-pattern", "_yyyyMMdd");
    intervalShardingAlgorithm.getProps().setProperty("datetime-interval-amount", Integer.toString(1));
    intervalShardingAlgorithm.getProps().setProperty("datetime-interval-unit", "DAYS");
    intervalShardingAlgorithm.init();
    List<String> availableTables = LongStream.range(0, ChronoUnit.DAYS.between(startTime, endTime))
            .mapToObj(startTime::plusDays)
            .map(localDate -> "t_order_" + localDate.format(DateTimeFormatter.ofPattern("yyyyMMdd")))
            .collect(Collectors.toList());
    Collection<String> actual = intervalShardingAlgorithm.doSharding(availableTables,
            new RangeShardingValue<>("t_order", "create_time",
                    new DataNodeInfo("t_order_", 54, '0'),
                    Range.closed(LocalDateTime.of(2022,3,11,2,25,27),
                            LocalDateTime.of(2022,3,12,2,25,27))));
    assertThat(actual.size(), is(2));
}
```

### Actual behavior

- Unfortunately it throws an inexplicable exception.
```shell
java.lang.StringIndexOutOfBoundsException: String index out of range: 23

	at java.lang.String.substring(String.java:1963)
	at org.apache.shardingsphere.sharding.algorithm.sharding.datetime.IntervalShardingAlgorithm.parseDateTime(IntervalShardingAlgorithm.java:155)
	at org.apache.shardingsphere.sharding.algorithm.sharding.datetime.IntervalShardingAlgorithm.hasIntersection(IntervalShardingAlgorithm.java:146)
	at org.apache.shardingsphere.sharding.algorithm.sharding.datetime.IntervalShardingAlgorithm.doSharding(IntervalShardingAlgorithm.java:137)
	at org.apache.shardingsphere.sharding.algorithm.sharding.datetime.IntervalShardingAlgorithm.doSharding(IntervalShardingAlgorithm.java:130)
	at org.apache.shardingsphere.sharding.algorithm.sharding.datetime.IntervalShardingAlgorithmTest.assertLocalDateTimeWithZero(IntervalShardingAlgorithmTest.java:215)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
```

### Reason analyze (If you can)

- The problem is with `org.apache.shardingsphere.sharding.algorithm.sharding.datetime.IntervalShardingAlgorithm#hasIntersection(final Range<LocalDateTime> calculateRange, final Range<Comparable<?>> range)` .
- image
- This method does not judge `LocalDateTime`, `Date`, `String` for `final Range<Comparable<?>> range`, so it executes `LocalDateTime.toString()` when `range.lowerEndpoint().toString()`, considering `LocalDateTime.parse(value.substring(0, dateTimePatternLength), dateTimeFormatter);` of `parseDateTime()`, which will cause `StringIndexOutOfBoundsException`.

- There is a simple unit test for LocalDateTime.toString() .
```java
@Test
void test2() {
    LocalDateTime parse = LocalDateTime.parse(
            "2022-02-16 00:00:00.000", DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS"));
    System.out.println("test---->" + parse);
}
```
- Execution yields the following result, 0 is automatically lost. This should be avoided.
```shell
test---->2022-02-16T00:00
```

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

- Execute unit tests directly.

### Example codes for reproduce this issue (such as a github link).

- Unit test is mentioned above.

### PR DESCRIPTION
**Pull request by linghengqian**
Fixes #16738.

Changes proposed in this pull request:
- IntervalShardingAlgorithm formats for LocalDateTime and Date.
- Add the corresponding unit tests.
- This is a simple fix, maybe with better handling to improve the performance of `IntervalShardingAlgorithm`.

### CODE REVIEW

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/IntervalShardingAlgorithm.java [hunk 1]
```diff
@@ -143,8 +143,20 @@ public String doSharding(final Collection<String> availableTargetNames, final Pr
...
+        if (range.hasLowerBound()) {
+            Comparable<?> lowerEndpoint = range.lowerEndpoint();
+            lower = lowerEndpoint instanceof LocalDateTime ? parseDateTime(((LocalDateTime) lowerEndpoint).format(dateTimeFormatter)) : parseDateTime(lowerEndpoint.toString());
+        } else {
+            lower = dateTimeLower;
+        }
+        LocalDateTime upper;
+        if (range.hasUpperBound()) {
+            Comparable<?> upperEndpoint = range.upperEndpoint();
+            upper = upperEndpoint instanceof LocalDateTime ? parseDateTime(((LocalDateTime) upperEndpoint).format(dateTimeFormatter)) : parseDateTime(upperEndpoint.toString());
```
Comment (zjcnb) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/IntervalShardingAlgorithm.java:156]: @linghengqian Thank for your pull request, but i think it should consider `Date` type

**Comment by linghengqian** [comment]
This PR was turned off by me until a new ONE came along to fix it.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/IntervalShardingAlgorithm.java (hunk 1 above)
Comment (linghengqian) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/IntervalShardingAlgorithm.java]: Maybe solved. There doesn't seem to be a `Util` class that lets me use pattern matching like JDK 17.

**Comment by linghengqian** [comment]
It looks like there is something wrong with my IntelliJ IDEA 2022.1, it swallows tabs with blank lines when submitting a commit.🤔

**Review by terrymanu — APPROVED** [review]

**Pull request merged**